File: lib_common/src/main/java/com/wujia/lib_common/data/network/DownSubscriber.java
package com.wujia.lib_common.data.network;

import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

/**
 * Created by xmren on 2018/3/9.
 */

public class DownSubscriber<ResponseBody> implements Subscriber<ResponseBody> {
    DownloadCallBack callBack;
    String saveFilePath;

    public DownSubscriber(String saveFilePath, DownloadCallBack callBack) {
        this.saveFilePath = saveFilePath;
        this.callBack = callBack;
    }


    @Override
    public void onError(Throwable e) {
        if (callBack != null) {
            callBack.onError(e);
        }
    }

    @Override
    public void onComplete() {
        if (callBack != null) {
            callBack.onCompleted();
        }
    }

    @Override
    public void onSubscribe(Subscription s) {
        if (callBack != null) {
            callBack.onStart();
        }
    }

    @Override
    public void onNext(ResponseBody responseBody) {
        DownLoadManager.getInstance(callBack).writeResponseBodyToDisk(saveFilePath, (okhttp3.ResponseBody) responseBody);
    }
}

File: lib_common/src/main/java/com/wujia/lib_common/data/network/DownSubscriber.kt
package com.wujia.lib_common.data.network

import org.reactivestreams.Subscriber
import org.reactivestreams.Subscription

/**
 * Created by xmren on 2018/3/9.
 */

class DownSubscriber<ResponseBody>(internal var saveFilePath: String, internal var callBack: DownloadCallBack?) : Subscriber<ResponseBody> {


    override fun onError(e: Throwable) {
        if (callBack != null) {
            callBack!!.onError(e)
        }
    }

    override fun onComplete() {
        if (callBack != null) {
            callBack!!.onCompleted()
        }
    }

    override fun onSubscribe(s: Subscription) {
        if (callBack != null) {
            callBack!!.onStart()
        }
    }

    override fun onNext(responseBody: ResponseBody) {
        DownLoadManager.getInstance(callBack).writeResponseBodyToDisk(saveFilePath, responseBody as okhttp3.ResponseBody)
    }
}
